File: sdk/ts/src/fsm.test.ts
import { test, expect } from "bun:test";
import { createMachine, type FSMDef, type Machine } from "./fsm";
import { newLogger } from "./log";
import type { Ctx } from "./thrall";
import type { Envelope } from "./envelope";

// A minimal ctx: the FSM handlers in these tests do not touch NATS.
function testCtx(): Ctx {
  return {
    nats: {} as unknown as Ctx["nats"],
    name: "t",
    app: "test",
    log: newLogger({ component: "test" }, { level: "error", format: "json", write: () => {} }),
    trace: "",
    call: (() => Promise.resolve(undefined)) as Ctx["call"],
    cast: () => {},
    startChild: async () => "",
    stopChild: async () => {},
  };
}

// turnstile mirrors the Go test machine: locked <-> unlocked, data counts pushes; "unlocked"
// has a coin reaction whose guard always rejects, and a "boom" that throws.
function turnstile(): FSMDef<number> {
  return {
    name: "turnstile",
    initial: "locked",
    init: () => 0,
    states: {
      locked: {
        on: { coin: { fn: (_e, d) => ({ next: "unlocked", data: d }) } },
      },
      unlocked: {
        on: {
          push: { fn: (_e, d) => ({ next: "locked", data: d + 1, reply: d + 1 }) },
          coin: { guard: () => false, fn: (_e, d) => ({ data: d }) },
          boom: {
            fn: () => {
              throw new Error("boom");
            },
          },
        },
      },
    },
  };
}

async function machineFor(def: FSMDef<number>): Promise<Machine<number>> {
  const ctx = testCtx();
  return createMachine(def, ctx, await Promise.resolve(def.init(ctx)), ctx.log);
}

async function callOp(m: Machine<number>, op: string): Promise<Envelope> {
  let reply: Envelope = { v: 1, kind: "reply" };
  await m.send({ op, payload: {}, kind: "call" }, { v: 1, kind: "call", op }, (r) => {
    reply = r;
  });
  return reply;
}

async function castOp(m: Machine<number>, op: string): Promise<void> {
  await m.send({ op, payload: {}, kind: "cast" }, { v: 1, kind: "cast", op });
}

async function waitState(m: Machine<number>, want: string, timeoutMs: number): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (m.state() === want) return;
    await new Promise((r) => setTimeout(r, 5));
  }
  throw new Error(`timeout waiting for state ${want}, still in ${m.state()}`);
}

test("starts in the initial state", async () => {
  const m = await machineFor(turnstile());
  expect(m.state()).toBe("locked");
});

test("transitions and replies on a call", async () => {
  const m = await machineFor(turnstile());
  const coin = await callOp(m, "coin");
  expect(coin.status).toBe("ok");
  expect(m.state()).toBe("unlocked");

  const push = await callOp(m, "push");
  expect(push.status).toBe("ok");
  expect(push.payload).toBe(1);
  expect(m.state()).toBe("locked");
  expect(m.data()).toBe(1);
});

test("unhandled event errors with no_transition and keeps state", async () => {
  const m = await machineFor(turnstile());
  const reply = await callOp(m, "push"); // no push in locked
  expect(reply.status).toBe("error");
  expect(reply.error?.type).toBe("no_transition");
  expect(m.state()).toBe("locked");
});

test("guard rejects the transition", async () => {
  const m = await machineFor(turnstile());
  await callOp(m, "coin"); // -> unlocked
  const reply = await callOp(m, "coin"); // guard false
  expect(reply.status).toBe("error");
  expect(reply.error?.type).toBe("guard_rejected");
  expect(m.state()).toBe("unlocked");
});

test("handler error replies with handler_error", async () => {
  const m = await machineFor(turnstile());
  await callOp(m, "coin");
  const reply = await callOp(m, "boom");
  expect(reply.status).toBe("error");
  expect(reply.error?.type).toBe("handler_error");
});

test("reserved _state op returns the current state", async () => {
  const m = await machineFor(turnstile());
  await callOp(m, "coin");
  const reply = await callOp(m, "_state");
  expect(reply.status).toBe("ok");
  expect((reply.payload as { state: string }).state).toBe("unlocked");
});

test("cast transitions without a reply", async () => {
  const m = await machineFor(turnstile());
  await castOp(m, "coin");
  expect(m.state()).toBe("unlocked");
});

test("send resolves only after the event is processed (durable ack-after-process contract)", async () => {
  // The durable consumer awaits onCast (= machine.send) before acking; that promise must not
  // resolve until the event has actually been handled, or a crash after ack loses the message.
  const m = await machineFor(turnstile());
  await m.send({ op: "coin", payload: {}, kind: "cast" }, { v: 1, kind: "cast", op: "coin" });
  expect(m.state()).toBe("unlocked");
});

test("drain resolves after in-flight events complete", async () => {
  const m = await machineFor(turnstile());
  void m.send({ op: "coin", payload: {}, kind: "cast" }, { v: 1, kind: "cast", op: "coin" });
  await m.drain();
  expect(m.state()).toBe("unlocked");
});

// timeoutMachine: "waiting" auto-transitions to "expired" after a state timeout, unless "ping"
// moves it to "active" (no timeout) first.
function timeoutMachine(afterMs: number): FSMDef<number> {
  return {
    name: "tmo",
    initial: "waiting",
    init: () => 0,
    states: {
      waiting: {
        on: {
          ping: { fn: (_e, d) => ({ next: "active", data: d }) },
          tick: { fn: (_e, d) => ({ next: "expired", data: d + 1 }) },
        },
        timeout: { after: afterMs, op: "tick" },
      },
      active: { on: {} },
      expired: { on: {} },
    },
  };
}

test("state timeout fires a transition", async () => {
  const m = await machineFor(timeoutMachine(30));
  await waitState(m, "expired", 1000);
  expect(m.data()).toBe(1);
});

test("an event cancels the state timeout", async () => {
  const m = await machineFor(timeoutMachine(30));
  await castOp(m, "ping"); // -> active before 30ms
  expect(m.state()).toBe("active");
  await new Promise((r) => setTimeout(r, 80));
  expect(m.state()).toBe("active"); // timeout did not fire after leaving the state
});

test("Outcome.timeout re-arms while staying", async () => {
  const def: FSMDef<number> = {
    name: "rearm",
    initial: "idle",
    init: () => 0,
    states: {
      idle: {
        on: {
          poke: { fn: (_e, d) => ({ data: d + 1, timeout: { after: 30, op: "done" } }) },
          done: { fn: (_e, d) => ({ next: "finished", data: d }) },
        },
      },
      finished: { on: {} },
    },
  };
  const m = await machineFor(def);
  await castOp(m, "poke");
  expect(m.state()).toBe("idle");
  await waitState(m, "finished", 1000);
  expect(m.data()).toBe(1);
});

File: sdk/python/fsm_test.py
"""FSM behaviour tests for the Python SDK - exercise the serialized machine core directly
(no running NATS), mirroring the Go and TS tests.

Run: uv run --with nats-py -m unittest fsm_test   (from sdk/python)
"""

import asyncio
import contextlib
import os
import unittest

from aether import (
    Ctx,
    Event,
    FSM,
    Logger,
    Outcome,
    Reaction,
    State,
    StateTimeout,
    ThrallDef,
    _Machine,
    start,
    start_fsm,
)


@contextlib.contextmanager
def _fake_env():
    """Set the env a thrall needs so the startup guards run, then restore. The URL is never
    dialled: every guard exercised here fires before nats.connect."""
    saved = {k: os.environ.get(k) for k in ("AETHER_NATS_URL", "AETHER_APP")}
    os.environ["AETHER_NATS_URL"] = "nats://127.0.0.1:1"
    os.environ["AETHER_APP"] = "test"
    try:
        yield
    finally:
        for k, v in saved.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v


def _boom(ev, d, ctx):
    raise RuntimeError("boom")


def turnstile() -> FSM:
    return FSM(
        name="turnstile",
        initial="locked",
        init=lambda ctx: 0,
        states={
            "locked": State(on={"coin": Reaction(fn=lambda ev, d, ctx: Outcome(next="unlocked", data=d))}),
            "unlocked": State(on={
                "push": Reaction(fn=lambda ev, d, ctx: Outcome(next="locked", data=d + 1, reply=d + 1)),
                "coin": Reaction(guard=lambda d, ev: False, fn=lambda ev, d, ctx: Outcome(data=d)),
                "boom": Reaction(fn=_boom),
            }),
        },
    )


def make(defn: FSM) -> _Machine:
    # Quiet logger so transition/event logs do not clutter test output.
    log = Logger({"component": "test"}, level=40, fmt="json", write=lambda line: None)
    ctx = Ctx(nats=None, name="t", app="test", log=log)
    return _Machine(defn, ctx, defn.init(ctx), ctx.log)


async def call_op(m: _Machine, op: str) -> dict:
    box: dict = {}
    await m.send(Event(op=op, payload={}, kind="call"), {"op": op}, lambda reply: box.__setitem__("r", reply))
    return box.get("r")


async def cast_op(m: _Machine, op: str) -> None:
    await m.send(Event(op=op, payload={}, kind="cast"), {"op": op}, None)


async def wait_state(m: _Machine, want: str, timeout: float = 1.0) -> None:
    loop = asyncio.get_running_loop()
    deadline = loop.time() + timeout
    while loop.time() < deadline:
        if m.state() == want:
            return
        await asyncio.sleep(0.005)
    raise AssertionError(f"timeout waiting for state {want}, still in {m.state()}")


class FSMTest(unittest.IsolatedAsyncioTestCase):
    async def test_starts_in_initial_state(self):
        m = make(turnstile())
        m.arm_initial()
        self.assertEqual(m.state(), "locked")

    async def test_transition_and_reply_on_call(self):
        m = make(turnstile())
        m.arm_initial()
        coin = await call_op(m, "coin")
        self.assertEqual(coin["status"], "ok")
        self.assertEqual(m.state(), "unlocked")

        push = await call_op(m, "push")
        self.assertEqual(push["status"], "ok")
        self.assertEqual(push["payload"], 1)
        self.assertEqual(m.state(), "locked")
        self.assertEqual(m.data, 1)

    async def test_unhandled_call_errors(self):
        m = make(turnstile())
        m.arm_initial()
        reply = await call_op(m, "push")  # no push in locked
        self.assertEqual(reply["status"], "error")
        self.assertEqual(reply["error"]["type"], "no_transition")
        self.assertEqual(m.state(), "locked")

    async def test_guard_rejects_transition(self):
        m = make(turnstile())
        m.arm_initial()
        await call_op(m, "coin")  # -> unlocked
        reply = await call_op(m, "coin")  # guard false
        self.assertEqual(reply["status"], "error")
        self.assertEqual(reply["error"]["type"], "guard_rejected")
        self.assertEqual(m.state(), "unlocked")

    async def test_handler_error_replies(self):
        m = make(turnstile())
        m.arm_initial()
        await call_op(m, "coin")
        reply = await call_op(m, "boom")
        self.assertEqual(reply["status"], "error")
        self.assertEqual(reply["error"]["type"], "handler_error")

    async def test_reserved_state_op(self):
        m = make(turnstile())
        m.arm_initial()
        await call_op(m, "coin")
        reply = await call_op(m, "_state")
        self.assertEqual(reply["status"], "ok")
        self.assertEqual(reply["payload"]["state"], "unlocked")

    async def test_cast_transitions_without_reply(self):
        m = make(turnstile())
        m.arm_initial()
        await cast_op(m, "coin")
        self.assertEqual(m.state(), "unlocked")


def timeout_machine(after: float) -> FSM:
    return FSM(
        name="tmo",
        initial="waiting",
        init=lambda ctx: 0,
        states={
            "waiting": State(
                on={
                    "ping": Reaction(fn=lambda ev, d, ctx: Outcome(next="active", data=d)),
                    "tick": Reaction(fn=lambda ev, d, ctx: Outcome(next="expired", data=d + 1)),
                },
                timeout=StateTimeout(after=after, op="tick"),
            ),
            "active": State(on={}),
            "expired": State(on={}),
        },
    )


class FSMTimeoutTest(unittest.IsolatedAsyncioTestCase):
    async def test_state_timeout_fires_transition(self):
        m = make(timeout_machine(0.03))
        m.arm_initial()
        await wait_state(m, "expired", 1.0)
        self.assertEqual(m.data, 1)

    async def test_event_cancels_state_timeout(self):
        m = make(timeout_machine(0.03))
        m.arm_initial()
        await cast_op(m, "ping")  # -> active before 30ms
        self.assertEqual(m.state(), "active")
        await asyncio.sleep(0.08)
        self.assertEqual(m.state(), "active")  # timeout did not fire after leaving

    async def test_outcome_rearms_timeout_while_staying(self):
        defn = FSM(
            name="rearm",
            initial="idle",
            init=lambda ctx: 0,
            states={
                "idle": State(on={
                    "poke": Reaction(fn=lambda ev, d, ctx: Outcome(data=d + 1, timeout=StateTimeout(after=0.03, op="done"))),
                    "done": Reaction(fn=lambda ev, d, ctx: Outcome(next="finished", data=d)),
                }),
                "finished": State(on={}),
            },
        )
        m = make(defn)
        m.arm_initial()
        await cast_op(m, "poke")
        self.assertEqual(m.state(), "idle")
        await wait_state(m, "finished", 1.0)
        self.assertEqual(m.data, 1)


class FSMHardeningTest(unittest.IsolatedAsyncioTestCase):
    async def test_enter_unknown_state_warns(self):
        lines: list = []
        log = Logger({"component": "test"}, level=30, fmt="json", write=lambda line: lines.append(line))
        ctx = Ctx(nats=None, name="t", app="test", log=log)
        defn = FSM(
            name="ghosted",
            initial="here",
            init=lambda ctx: 0,
            # "ghost" is not a defined state
            states={"here": State(on={"leave": Reaction(fn=lambda ev, d, ctx: Outcome(next="ghost", data=d))})},
        )
        m = _Machine(defn, ctx, defn.init(ctx), ctx.log)
        m.arm_initial()
        await cast_op(m, "leave")
        self.assertTrue(any("fsm transition to unknown state" in line for line in lines), lines)

    async def test_start_fsm_rejects_unknown_initial(self):
        defn = FSM(name="typo", initial="startt", init=lambda ctx: 0, states={"start": State(on={})})
        with _fake_env(), self.assertRaises(RuntimeError) as cm:
            await start_fsm(defn)
        self.assertIn("not in states", str(cm.exception))

    async def test_start_fsm_rejects_missing_init(self):
        defn = FSM(name="no-init", initial="a", init=None, states={"a": State(on={})})
        with _fake_env(), self.assertRaises(RuntimeError) as cm:
            await start_fsm(defn)
        self.assertIn("init is required", str(cm.exception))

    async def test_start_rejects_missing_init(self):
        defn = ThrallDef(name="no-init", init=None)
        with _fake_env(), self.assertRaises(RuntimeError) as cm:
            await start(defn)
        self.assertIn("init is required", str(cm.exception))


if __name__ == "__main__":
    unittest.main()
